### PR TITLE
feat: Implement filter for projects

### PR DESCRIPTION
**Pull request by david-code**
## Description

- This PR creates a new `ListFilter` component, that can filter a list of objects by text inputs, and by select inputs. These filters are defined by the props passed to the component. The code is honestly a bit too complex, and I'm sure it can be simplified. The good thing is I also implemented some unit tests, so we should be able to refactor this component without being too afraid of breaking things.

- The test setup went pretty smoothly, but there are still a few rough patches. I left a bunch of TODO comments with some things I would ideally like to clean up in the future. I think I'll start tackling them when we start writing tests for new components.

- Implements the ListFilter for the Project List. Eventually it will be used for SiteMap / LocationDropdown and Site Project List, but this PR is big enough as it is. We'll tackle these issues in future PRs as well.

### Checklist
- [x] Corresponding issue has been opened
- [x] New tests added

### Related Issues
Fixes https://github.com/techmatters/terraso-mobile-client/issues/143

### CODE REVIEW

File: dev-client/jest/setup.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import '@testing-library/jest-native';
```
Comment (paulschreiber) [inline dev-client/jest/setup.ts:1]: issue: add copyright notice.

File: dev-client/__tests__/components/common/ListFilter.test.tsx [hunk 2]
```diff
@@ -0,0 +1,274 @@
+import {render, screen, fireEvent} from '@testing-library/react-native';
```
Comment (paulschreiber) [inline dev-client/__tests__/components/common/ListFilter.test.tsx:1]: issue: add copyright notice.

File: dev-client/src/components/common/EmbadgedIcon.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import {Badge, Pressable} from 'native-base';
```
Comment (paulschreiber) [inline dev-client/src/components/common/EmbadgedIcon.tsx:1]: issue: add copyright notice.

File: dev-client/src/components/common/ListFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,278 @@
+import {Button, FormControl, HStack, Input, Select, VStack} from 'native-base';
```
Comment (paulschreiber) [inline dev-client/src/components/common/ListFilter.tsx:1]: issue: add copyright notice.

File: dev-client/src/screens/ProjectListScreen.tsx [hunk 5]
```diff
@@ -69,23 +73,51 @@ export const ProjectListScreen = () => {
...
+                label: 'We will getrid of this',
+              },
+              select: {
+                role: {
+                  label: 'Filter projects by role',
+                  placeholder: '',
+                  options: {
+                    manager: 'Manager',
+                    contributor: 'Contributor',
+                    viewer: 'Viewer',
```
Comment (paulschreiber) [inline dev-client/src/screens/ProjectListScreen.tsx:101]: issue: localize these strings.

File: dev-client/src/components/common/ListFilter.tsx [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        trigger={onOpen => (
+          <HStack space="20px" mb="15px">
+            <EmbadgedIcon
+              iconName="filter-list"
+              onPress={onOpen}
+              badgeNum={numFilters}
+              accessibilityLabel="Update list filters"
+              _badge={{
+                accessibilityLabel: 'Filters applied',
+              }}
```
Comment (paulschreiber) [inline dev-client/src/components/common/ListFilter.tsx:236]: issue: localize these strings.

File: dev-client/src/components/common/ListFilter.tsx [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+              bg="background.default"
+              pl="8px"
+              py="6px"
+              InputLeftElement={<Icon ml="16px" name="search" />}
+              InputRightElement={
+                query.length === 0 ? undefined : (
+                  <IconButton
+                    name="close"
+                    onPress={clearTextInputFilter}
+                    accessibilityLabel="Clear search filter"
```
Comment (paulschreiber) [inline dev-client/src/components/common/ListFilter.tsx:255]: issue: localize this string.

File: dev-client/jest/setup.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+  return Reanimated;
+});
+
+// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
+jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
+//jest.mock(
+//'native-base/lib/commonjs/components/composites/Transitions/Transition',
+//  () => 'Transition',
+//);
```
Comment (paulschreiber) [inline dev-client/jest/setup.ts:22]: issue: remove commented-out code.

File: dev-client/src/components/common/ListFilter.tsx [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    InputFilter: React.ReactNode;
+  }) => React.ReactNode;
+};
+
+const normalizeText = (text: string) =>
+  text
+    .trim()
+    .toLowerCase()
+    .normalize('NFD')
+    .replace(/[\u0300-\u036f]/g, ''); // unicode range for combining diacritical marks
```
Comment (paulschreiber) [inline dev-client/src/components/common/ListFilter.tsx:69]: issue: this should be a global helper function, not specific to this component. Also: we do something similar in web client. See if that can be reused.

**Review by paulschreiber — CHANGES_REQUESTED** [review]

File: dev-client/jest/setup.ts (hunk 1 above)
Comment (david-code) [reply dev-client/jest/setup.ts:1]: @paulschreiber I noticed only one file has this notice right now. Let's use [eslint-plugin-header](https://www.npmjs.com/package/eslint-plugin-header) to catch this and add the copyright automatically for the other files in another PR.

File: dev-client/package.json [hunk 10]
```diff
@@ -71,6 +73,17 @@
...
   "jest": {
-    "preset": "react-native"
+    "preset": "react-native",
+    "setupFilesAfterEnv": [
+      "@testing-library/jest-native/extend-expect",
+      "./node_modules/react-native-mmkv-storage/jest/mmkvJestSetup.js",
+      "<rootDir>/jest/setup.ts"
+    ],
+    "moduleNameMapper": {
+      "\\.(jpg|ico|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.ts"
```
Comment (paulschreiber) [inline dev-client/package.json:83]: question: where does this list of extensions come from?

File: dev-client/src/util.ts [hunk 11]
```diff
@@ -25,3 +25,10 @@ export const removeKeys = (a: any, b: any) => {
     }
   }
 };
+
+export const normalizeText = (text: string) =>
```
Comment (paulschreiber) [inline dev-client/src/util.ts:29]: this is a duplicate of `cleanSensitiveCharacters` in the web client. Please refactor and move to shared client.

**Review by paulschreiber — CHANGES_REQUESTED** [review]

**Comment by david-code** [comment]
superseded by #633

**Pull request closed without merging**